iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -139,8 +144,328 @@ static Value flatten(PatternRewriter &rewriter, Location loc, Value vector) {
   return rewriter.create<vector::ShapeCastOp>(loc, dstType, vector);
 }
 
-/// Converts matrix-times-matrix-transposed vector.contracts with
-/// lhs and rhs inputs defined by arith.extsi promoting from i8 to i32,
+// Asserts that i is a power of two, and returns its log2.
+// Note: the llvm helpers used internally operate on uint32, but we keep that
+// an internal detail as the surrounding code here is all operating on signed
+// integers and mixing signed and unsigned would be error-prone.
+int8_t exactLog2(int32_t i) {
+  assert(i > 0);
+  uint32_t u = i;
+  assert(llvm::isPowerOf2_32(u));
+  return llvm::countTrailingZeros(u);
+}
+
+// Helper to handle powers of two size computations without the overhead
+// of runtime divisions. Divisions remain very expensive compared to most other
+// instructions.  Divisions are of course cheap when the divisor is
+// a constant, but a typical use case for us is
+//
+//    lhsBitWidth / kernel.registerBitWidth
+//
+// kernel.registerBitWidth is *initialized* from a literal value (say 128) but
```

Comment:
Depends what we mean by *throughout* and what benefit we want to achieve with `constexpr` here.

As said in reply to Mahesh, to avoid binary code duplication, we want only a single `Pattern` to be built and fed the different kernels at runtime, as opposed to compile-time specializing the `Pattern` code to each kernel.

So it actually matches intent that the `Kernel` objects here are *used* as runtime data.

Given that, they could still be `constexpr`, but that would not really make a difference to the generated code. That would also be kind of annoying in C++14 (what would be the type of `implAsm`? A pointer can't be `constexpr`, and it's only in C++17 that `std::string` becomes `constexpr`).